<a href="https://colab.research.google.com/github/Gaurav-DSE/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*


I chose the **Refresh / Content Opportunity Scoring** lane.

Many websites contain hundreds or thousands of pages, but content teams have limited time and resources. Instead of reviewing every page manually, a scoring system can help prioritize which pages should be reviewed first.

This lane interests me because it combines business decision-making with machine learning. The goal is not just to train a model, but to  focus on pages that may benefit the most from review, refresh, or optimization.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

### Search Question

Which webpages should be reviewed first for content refresh or optimization?

### Unit of Analysis

One webpage (one content page).

### Output

A ranked review queue or priority score for each webpage.

### Decision

Which pages should be selected for review first.

### Action

review selected pages and decide whether to:
- Update outdated information
- Improve titles or descriptions
- Expand content
- Improve engagement signals
- Monitor performance more closely

### Cost of a Wrong Recommendation

False Positive:
A page is recommended for review even though it does not need attention, causing time and effort to be wasted.

False Negative:
An important page that needs attention is missed, potentially allowing traffic, engagement, or visibility to continue declining.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [9]:

import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
elif os.path.basename(os.getcwd()) == "notebooks":
    os.chdir("..")

import pandas as pd, numpy as np
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

In [10]:
df.columns

Index(['content_id', 'client_id', 'search_volume', 'competition',
       'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count',
       'char_count', 'provider_used', 'model_used', 'impressions_90d',
       'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d',
       'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d',
       'days_with_impressions', 'days_with_sessions', 'impressions_last_30d',
       'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d',
       'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier',
       'age_tier_order', 'days_since_last_update', 'freshness_tier',
       'word_count_tier', 'char_count_tier', 'ctr', 'avg_position',
       'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier',
       'position_tier', 'trend_direction', 'trend_pct'],
      dtype='object')

In [11]:
print("Total pages:", len(df))

print(
    "Declining pages (%):",
    round((df["trend_direction"].str.lower() == "down").mean() * 100, 2))

print(
    "Average days since last update:",
    round(df["days_since_last_update"].mean(), 1))

print(
    "Average CTR (%):",
    round(df["ctr"].mean() * 100, 2))

print("Average content age (days):",
      round(df["content_age_days"].mean(), 1))


Total pages: 30000
Declining pages (%): 54.21
Average days since last update: 46.1
Average CTR (%): 51.07
Average content age (days): 256.2


## Quick Look at the Data

The starter dataset contains 30,000 webpages.

About 54.21% of pages show a declining trend direction, suggesting that a large portion of content may require intervention.

Pages have not been updated recently on average, with a mean of 46.1 days since the last update. The average content age is 256.2 days, indicating that many pages may become stale over time.

The average CTR is 51.07%, showing that user engagement varies across pages and may present opportunities for optimization.

These observations support the Refresh / Content Opportunity Scoring lane because teams cannot manually inspect thousands of pages. A ranked review queue could help prioritize which pages should be reviewed first for refresh, optimization, or monitoring.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

This project aims to identify webpages that may deserve review based on
observed content, traffic, engagement, and freshness signals. The recommendations are intended to support decision-making rather than replace human judgment.

A high priority score does not guarantee that refreshing a page will improve its performance. The analysis identifies patterns and associations in the data, but it does not prove causation. External factors such as seasonality, competition, search engine updates, and user behavior may also affect page performance.

The final output should be treated as a ranked review queue that helps teams focus their attention more efficiently.

## Self-check

Before you submit, confirm each line honestly:

- [X] Every section above is filled — markdown thinking AND the code that backs it
- [X] The notebook runs top to bottom with no errors (Runtime → Run all)
- [X] No client names, URLs, or private queries anywhere
- [X] My claims use careful words: observed, measured, directional, decision-support
- [X] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.